Fireproofing implementation refactor

samsymons opened the pull request:
Task/Issue URL: https://app.asana.com/0/414709148257752/1199984099899045
Tech Design URL:
CC:

**Description**:

This PR updates the Fireproofing implementation to remove the persistence step during the Fire button process. The new implementation deletes all cached data except for cookies, and then iterates over the cookies to delete those which aren't Fireproof.

**Steps to test this PR**:
1. Test that Fireproofing works

<!--
Before submitting a PR, please ensure you have tested the combinations you expect the reviewer to test, then delete configurations you *know* do not need explicit testing.

Using a simulator where a physical device is unavailable is acceptable. 
-->

**Device Testing**:

* [x] iPhone SE (1st Gen)
* [x] iPhone 8
* [x] iPhone X
* [x] iPad

**OS Testing**:

* [x] iOS 11
* [x] iOS 12
* [x] iOS 13
* [x] iOS 14

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**

samsymons committed:
Remove the flag for consuming cookies.

samsymons committed:
Remove all cookies except those from Fireproof domains.

samsymons committed:
Disable a SwiftLint warning, we were right at the threshold.

bwaresiak commented on Core/CookieStorage.swift:
```diff
@@ -28,57 +28,13 @@ public class CookieStorage {
+    public func clear() {
         userDefaults.removeObject(forKey: Constants.key)
         os_log("cleared cookies", log: generalLog, type: .debug)
     }
```
I'm all for removing the code - but shouldn't we try to migrate in case cookies are backed in the user defaults?

brindy replied:
So something occurs to me:

1) Migration is unnecessary because we always have a webview, OR 
2) We shouldn't be recreating the webview so quickly after the fire button is used (put a breakpoint on TabViewController.swift:303 and hit the fire button)

bwaresiak replied:
There is a slight chance cookies list this won't be empty, exactly due to #2 you raised: if cookies were stored with delay and after WebView has been created. It is unlikely, but I still wonder if we should try to preserve those in such case...

samsymons replied:
Fair point @bwaresiak, there's certainly a slim chance where cookies have been stored but not yet consumed. I've restored the `consumeCookies` logic to ensure that any stored cookies are being restored, the same way we've always been doing. Some time in the future, we can remove it again.

brindy replied:
I just think it's so edge case it's not worth having it in the code base.  I think we'd be better off addressing the issue that there's a webview constructed during the process of deleting data.

bwaresiak replied:
I think in such case we should value potential consistency over cleaning up codebase - we can still remove this migration as a follow up task scheduled few months from now. 

@brindy WebView construction seem to be a separate issue: as far as I see, this is about when to construct WebView for newly created tabs (and cleaning data's side effect is recreating first tab). I think we opted in to do that immediately, as doing it in a lazy manner can result is noticeable lag on first navigation - still, I'm more than happy to discuss that in Asana if you think we should revisit that and it should happen as a separate project.

brindy replied:
Yeah, you're right, let's clean it up in a few months.

bwaresiak commented on Core/WebCacheManager.swift:
```diff
@@ -101,62 +64,60 @@ public class WebCacheManager {
+            }
+
+            DispatchQueue.global(qos: .background).async {
+                _ = group.wait(timeout: .now() + 5)
```
Maybe we should fire debug pixel when it's timing out?

bwaresiak replied:
Another question: is it possible that this executes before 'getAllCookies' returns (and before first 'enter' is called on a group)?

brindy replied:
Yeah. probably want an `enter()` before calling `getAllCookies` and a `leave()` inside it, assuming that the completion handler for `getAllCookies` always gets called.  If it doesn't ... not sure what you could do here.

bwaresiak replied:
We also could move whole `DispatchQueue.global(qos: .background).async` and nest it at the end of `getAllCookies` completion block - also assuming it is being always called. :)

samsymons replied:
I like that suggestion, moving the `DispatchGroup` handling into the `getAllCookies` callback sounds good  – done.

brindy reviewed (commented):
I see @bwaresiak has assigned it so I'll let him give the final approval, but I've been running it on my device for the last couple of days and it seems to be working great.

samsymons committed:
Fire a debug pixel if the call to WebCacheManager.clear times out.

samsymons committed:
Move the DispatchGroup code inside the getAllCookies call.

samsymons committed:
Restore the cookie consumption logic.

While this will go unused in 99% of cases, there is a chance that cookies have been persisted and not yet consumed, so the `consumeCookies` call should remain for now.

samsymons committed:
Fix weird indentation after restoring previous logic.

samsymons committed:
Fix a little more weird indentation. Not sure what Xcode’s up to here!

samsymons committed:
Xcode’s whitespace trimming seems to have stopped working.

samsymons committed:
Remove the `setCookie` calls, don’t need to restore those.

brindy commented on Core/CookieStorage.swift:
```diff
@@ -57,28 +57,4 @@ public class CookieStorage {
-        //  untouched for previous versions of iOS as changing from Strict to Lax blindly could be a security issue if sites have set it as such.
-        //  See https://www.owasp.org/index.php/SameSite
-        if #available(iOS 13, *) {
-            cookieData[HTTPCookiePropertyKey.sameSitePolicy.rawValue] = cookie.sameSitePolicy ?? HTTPCookieStringPolicy.sameSiteLax
```
So ... another problem here is that we make this change deliberately.  It's not quite the right logic though, so I'm just getting confirmation we actually need it

cc @kdzwinel

brindy replied:
I talked to Konrad about this and we're all good given that webkit blocks 3p cookies anyway (so long as the user hasn't disabled that, which we can't do much about)

bwaresiak commented on Core/WebCacheManager.swift:
```diff
@@ -101,62 +103,61 @@ public class WebCacheManager {
+                }
+
+                DispatchQueue.global(qos: .background).async {
+                    _ = group.wait(timeout: .now() + 5)
```
Wasn't we supposed to use `cookieDeletionTimedOut` here? :)

samsymons replied:
Correct pixel is being used here now 🙂

bwaresiak commented on Core/WebCacheManager.swift:
```diff
@@ -101,62 +103,61 @@ public class WebCacheManager {
+                DispatchQueue.global(qos: .background).async {
+                    _ = group.wait(timeout: .now() + 5)
+
+                    Pixel.fire(pixel: .blankOverlayNotDismissed)
```
And this piece of code is pretty surprising :)

samsymons replied:
Hmm wtf happened here 🤔

samsymons committed:
Use the correct pixel in the cache manager. Don’t code while sleep deprived, kids.

bwaresiak commented on Core/WebCacheManager.swift:
```diff
@@ -101,62 +103,61 @@ public class WebCacheManager {
+
+                DispatchQueue.global(qos: .background).async {
+                    _ = group.wait(timeout: .now() + 5)
+                    Pixel.fire(pixel: .cookieDeletionTimedOut)
```
But aren't we supposed to only fire that when there is a timeout? :)

samsymons replied:
D'oh, yes sorry, am not able to focus properly right now. 🙂 I'll change this to check against the timeout result.

samsymons replied:
Done

samsymons committed:
Fire the pixel when the wait call times out.

samsymons committed:
Add parameters to track the cache manager timeout.

bwaresiak reviewed (approved):
Good job @samsymons !
